components/profile/heartbeats.rs

```diff
@@ -11,86 +11,68 @@ use std::collections::HashMap;
 use std::env::var_os;
 use std::error::Error;
 use std::fs::File;
-use std::mem;
 use std::path::Path;
+use std::sync::atomic::{ATOMIC_BOOL_INIT, AtomicBool, Ordering};
 
 
 static mut HBS: Option<*mut HashMap<ProfilerCategory, Heartbeat>> = None;
 
+// unfortunately can't encompass the actual hashmap in a Mutex (Heartbeat isn't Send/Sync), so we'll use a spinlock
+static mut HBS_SPINLOCK: AtomicBool = ATOMIC_BOOL_INIT;
+
+fn lock_and_work<F, R>(work: F) -> R where F: FnOnce() -> R {
```

Comment:
To enforce that `HBS` is only used in a sound way you can move the declaration inside this function I believe (or in another module, so it's private to everything except this function), and give it as a parameter to `work`.